code/modules/admin/ipintel.dm

```diff
@@ -0,0 +1,224 @@
+	res.intel = ip_intel_query(ip)
+	if(updatecache && res.intel >= 0)
+		SSipintel.cache[ip] = res
+		if(dbcon.IsConnected())
+			var/DBQuery/query_add_ip_intel = dbcon.NewQuery("INSERT INTO [format_table_name("ipintel")] (ip, intel) VALUES (INET_ATON('[ip]'), [res.intel]) ON DUPLICATE KEY UPDATE intel = VALUES(intel), date = NOW()")
+			query_add_ip_intel.Execute()
+			qdel(query_add_ip_intel)
+
+
+/proc/ip_intel_query(ip, var/retryed=0)
+	. = -1 //default
+	if(!ip)
+		return
+	if(SSipintel.throttle > world.timeofday)
+		return
+	if(!SSipintel.enabled)
+		return
+
+	var/list/http[] = world.Export("http://[config.ipintel_domain]/check.php?ip=[ip]&contact=[config.ipintel_email]&format=json&flags=b")
+
+	if(http)
+		var/status = text2num(http["STATUS"])
+
+		if(status == 200)
+			var/response = json_decode(file2text(http["CONTENT"]))
+			if(response)
+				if(response["status"] == "success")
+					var/intelnum = text2num(response["result"])
+					if(isnum(intelnum))
+						return text2num(response["result"])
+					else
+						ipintel_handle_error("Bad intel from server: [response["result"]].", ip, retryed)
+						if(!retryed)
+							sleep(25)
+							return .(ip, 1)
+				else
+					ipintel_handle_error("Bad response from server: [response["status"]].", ip, retryed)
+					if(!retryed)
+						sleep(25)
+						return .(ip, 1)
+
+		else if(status == 429)
+			ipintel_handle_error("Error #429: We have exceeded the rate limit.", ip, 1)
+			return
+		else
+			ipintel_handle_error("Unknown status code: [status].", ip, retryed)
+			if(!retryed)
+				sleep(25)
+				return .(ip, 1)
+	else
+		ipintel_handle_error("Unable to connect to API.", ip, retryed)
+		if(!retryed)
+			sleep(25)
+			return .(ip, 1)
+
+
+/proc/ipintel_handle_error(error, ip, retryed)
+	if(retryed)
+		SSipintel.errors++
+		error += " Could not check [ip]. Disabling IPINTEL for [SSipintel.errors] minute[( SSipintel.errors == 1 ? "" : "s" )]"
+		SSipintel.throttle = world.timeofday + (10 * 120 * SSipintel.errors)
+	else
+		error += " Attempting retry on [ip]."
+	log_ipintel(error)
+
+/proc/log_ipintel(text)
+	log_game("IPINTEL: [text]")
+	log_debug("IPINTEL: [text]")
+
+
+/proc/ipintel_is_banned(t_ckey, t_ip)
+	if(!config.ipintel_email)
+		return FALSE
+	if(!config.ipintel_whitelist)
+		return FALSE
+	if(!dbcon.IsConnected())
+		return FALSE
+	if(!ipintel_badip_check(t_ip))
+		return FALSE
+	if(vpn_whitelist_check(t_ckey))
+		return FALSE
+	return TRUE
+
+/proc/ipintel_badip_check(target_ip)
+	var/rating_bad = config.ipintel_rating_bad
+	if(!rating_bad)
+		log_debug("ipintel_badip_check reports misconfigured rating_bad directive")
+		return FALSE
+	var/valid_hours = config.ipintel_save_bad
+	if(!valid_hours)
+		log_debug("ipintel_badip_check reports misconfigured ipintel_save_bad directive")
+		return FALSE
+	var/check_sql = {"SELECT * FROM [format_table_name("ipintel")] WHERE ip = INET_ATON('[target_ip]') AND intel >= [rating_bad] AND (date + INTERVAL [valid_hours] HOUR) > NOW()"}
```

Comment:
Reviewer's note:
If `config.ipintel_rating_bad` is modified through some means to an SQL injection string, this could be used for a DB attack. It would be wise to wrap the assignments in `sanitizeSQL`

Reply:
Changed